### PR TITLE
[Doc] Improve `thread::spawn` documentation

### PR DESCRIPTION
**Pull request by gamazeps**
Part of #29378

- Add two examples to `thread::spawn` doumentation that show common uses of threads.
- Add a link to `thread::Builder` in the `thread::spawn` documentation for configuring threads.
- Add a link to `thread::spawn` in `thread::Builder` in order to avoid documentation duplication.

r? @steveklabnik

### CODE REVIEW

File: src/libstd/thread/mod.rs [hunk 1]
```diff
@@ -409,6 +411,54 @@ impl Builder {
 ///
 /// handler.join().unwrap();
 /// ```
+///
+/// As mentionned in the module documentation, threads are usualy made to
+/// communicate using [`channel`s][`channels`], here is how it usually looks.
```
Comment (steveklabnik) [inline src/libstd/thread/mod.rs]: this should work without the second set of `[]`s by using the same text below

File: src/libstd/thread/mod.rs [hunk 2]
```diff
@@ -387,19 +389,19 @@ impl Builder {
...
 ///
-/// [`JoinHandle`]: ../../std/thread/struct.JoinHandle.html
-/// [`join`]: ../../std/thread/struct.JoinHandle.html#method.join
-/// [`Err`]: ../../std/result/enum.Result.html#variant.Err
-/// [`panic`]: ../../std/macro.panic.html
-/// [`Builder::spawn`]: ../../std/thread/struct.Builder.html#method.spawn
-///
 /// # Examples
 ///
+/// Simple thread creation.
```
Comment (steveklabnik) [inline src/libstd/thread/mod.rs]: Could we say

> Creating a thread:

instead here? I try to stick away from "simple" 😄

**Review by steveklabnik — COMMENTED** [review]
a few small changes please! Thanks for this, looks great!

**Comment by gamazeps** [comment]
Done :)

**Comment by steveklabnik** [comment]
```
[01:06:47] failures:
[01:06:47] 
[01:06:47] ---- thread/mod.rs - thread::spawn (line 445) stdout ----
[01:06:47] 	error[E0425]: cannot find value `v` in this scope
[01:06:47]   --> <anon>:12:16
[01:06:47]    |
[01:06:47] 12 | println!("{}", v);
[01:06:47]    |                ^ not found in this scope
[01:06:47] 
[01:06:47] error: aborting due to previous error(s)
[01:06:47] 
[01:06:47] thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc/session/mod.rs:217
[01:06:47] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:06:47] 
[01:06:47] ---- thread/mod.rs - thread::spawn (line 422) stdout ----
[01:06:47] 	error: unused result which must be used
[01:06:47]   --> <anon>:10:5
[01:06:47]    |
[01:06:47] 10 |     tx.send("Hello, thread".to_owned());
[01:06:47]    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[01:06:47]    |
[01:06:47]    = note: #[deny(unused_must_use)] implied by #[deny(warnings)]
[01:06:47] note: lint level defined here
[01:06:47]   --> <anon>:1:9
[01:06:47]    |
[01:06:47] 1  | #![deny(warnings)]
[01:06:47]    |         ^^^^^^^^
[01:06:47] 
[01:06:47] error: unused result which must be used
[01:06:47]   --> <anon>:17:1
[01:06:47]    |
[01:06:47] 17 | sender.join();
[01:06:47]    | ^^^^^^^^^^^^^^
[01:06:47]    |
[01:06:47]    = note: #[deny(unused_must_use)] implied by #[deny(warnings)]
[01:06:47] 
[01:06:47] error: unused result which must be used
[01:06:47]   --> <anon>:18:1
[01:06:47]    |
[01:06:47] 18 | receiver.join();
[01:06:47]    | ^^^^^^^^^^^^^^^^
[01:06:47]    |
[01:06:47]    = note: #[deny(unused_must_use)] implied by #[deny(warnings)]
[01:06:47] 
[01:06:47] error: aborting due to previous error(s)
[01:06:47] 
[01:06:47] thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc/session/mod.rs:217
[01:06:47] 
[01:06:47] 
[01:06:47] failures:
[01:06:47]     thread/mod.rs - thread::spawn (line 422)
[01:06:47]     thread/mod.rs - thread::spawn (line 445)
[01:06:47] 
[01:06:47] test result: FAILED. 797 passed; 2 failed; 22 ignored; 0 measured
[01:06:47] 
[01:06:47] error: test failed, to rerun pass '--doc'
[01:06:47] 
[01:06:47] 
[01:06:47] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "-j" "4" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--features" "panic-unwind jemalloc backtrace" "-p" "std:0.0.0" "-p" "alloc_system:0.0.0" "-p" "libc:0.0.0" "-p" "compiler_builtins:0.0.0" "-p" "rustc_asan:0.0.0" "-p" "rustc_lsan:0.0.0" "-p" "unwind:0.0.0" "-p" "core:0.0.0" "-p" "rand:0.0.0" "-p" "rustc_msan:0.0.0" "-p" "panic_abort:0.0.0" "-p" "alloc:0.0.0" "-p" "collections:0.0.0" "-p" "rustc_tsan:0.0.0" "-p" "std_unicode:0.0.0" "--"
[01:06:47] expected success, got: exit code: 101
[01:06:47] 
[01:06:47] 
[01:06:47] Build completed unsuccessfully in 0:27:07
[01:06:47] Makefile:54: recipe for target 'check' failed
[01:06:47] make: *** [check] Error 1
```

once this is fixed we should be good to go!

**Comment by gamazeps** [comment]
Done :)

File: src/libstd/thread/mod.rs [hunk 3]
```diff
@@ -409,6 +411,54 @@ impl Builder {
 ///
 /// handler.join().unwrap();
 /// ```
+///
+/// As mentionned in the module documentation, threads are usualy made to
```
Comment (parir) [inline src/libstd/thread/mod.rs]: s/mentionned/mentioned/
s/usualy/usually/

Comment (gamazeps) [inline src/libstd/thread/mod.rs]: Done :)

**Pull request merged**